### PR TITLE
Adds the "Jetpack powered" banner to the Notifications screen

### PR DESCRIPTION
**Pull request by Gio2018**
Fixes #NA

This PR adds the "Jetpack powered" banner to the Stats and Notifications screens. This change only affects the WordPress app, not the Jetpack app

On iPad and iPhone regular landscape (e.g. iPhone 13 Pro Max), the FAB will adjust its position if the banner is visible

iPad Notifications Dark | iPhone Notifications Light
-- | --
 | 

To test:
Test on iPhone, iPad and larger iPhone (e.g iPhone 13 Pro Max)
1.  Build/run this branch on the `WordPress` target
2. Tap on Notifications and make sure the banner is visible
3. Build the Jetpack target from this same branch, repeat steps 2 and 5 and make sure the banner does not appear in the Jetpack App

## Regression Notes
1. Potential unintended areas of impact
Low: this PR adds a new property to WPTabBarController, and make some runtime adjustments to view to account for the banner being visible/not visible

2. What I did to test those areas of impact (or what existing automated tests I relied on)
test views on different devices, quick check if there was something obviously wrong in the memory graph

3. What automated tests I added (or what prevented me from doing so)
None
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~ RN are frozen for this release

### CODE REVIEW

**Comment by guarani** [comment]
Thanks, @Gio2018! Reviewing now.

**Review by guarani — COMMENTED** [review]
Hi @Gio2018, after switching around between My Site, Reader, and Notifications, the banner gets stuck:

|   |  | 
| - | - |



Let me know if you have trouble reproducing. In the meantime I'll try to collect exact steps to reproduce.

**Comment by guarani** [comment]
On the Stats Days/Weeks/Months/Years tabs, the banner covers the last part of the screen "No data for this period" message:

https://user-images.githubusercontent.com/1898325/176959422-7b8fe5a4-d396-4931-b643-dcc4e0a24733.mov

File: WordPress/Classes/ViewRelated/Jetpack Branding/WPTabBarController+JetpackBanner.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    let rect = CGRect(origin: .zero, size: size)
+    ctx.setFillColor(color.cgColor)
+    ctx.fill(rect)
+    ctx.concatenate(CGAffineTransform(a: 1, b: 0, c: 0, d: -1, tx: 0, ty: size.height))
+    ctx.draw(image, in: rect)
+
+    return UIGraphicsGetImageFromCurrentImageContext() ?? self
+  }
+}
```
Comment (guarani) [inline WordPress/Classes/ViewRelated/Jetpack Branding/WPTabBarController+JetpackBanner.swift]: Is this used?

File: WordPress/Classes/ViewRelated/Jetpack Branding/WPTabBarController+JetpackBanner.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            view.clipsToBounds = true
+            NSLayoutConstraint.activate([
+                view.centerXAnchor.constraint(equalTo: imageView.centerXAnchor),
+                view.centerYAnchor.constraint(equalTo: imageView.centerYAnchor),
+                view.heightAnchor.constraint(equalToConstant: Appearance.jetpackIconBackgroundSize),
+                view.widthAnchor.constraint(equalToConstant: Appearance.jetpackIconBackgroundSize)
+            ])
+        }
+
+        return jetpackButton
```
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/Jetpack Branding/WPTabBarController+JetpackBanner.swift]: Is it intentional that you use an `UIButton`… but don't associate a target/action to it?

I didn't test on device in practice (only reading from the code), but wouldn't that make this tappable visually (as in, if you tap on the banner you're gonna have a visual feedback of the button being in a "pressed" state) while nothing would happen when tapped?

**Review by AliSoftware — COMMENTED** [review]
I have very little knowledge of the architecture of the app's UI and only reviewing the diff (because I'm away from my computer RN and couldn't test and check in Xcode), but I'm wondering if it wouldn't be simpler and also safer to instead:

 - Make a [Container ViewController](https://developer.apple.com/library/archive/featuredarticles/ViewControllerPGforiPhoneOS/ImplementingaContainerViewController.html) that would just have a `childViewController` on top, and the banner at the bottom, and appropriate layoutConstraints to keep them in lock
 - Then replace the ViewControllers that are currently associated with Notifications and Reader tabs with an instance of this `ContainerViewController`, containing the original VC as its `childViewController`

My thinking is that going this route would avoid you to have to observe notifications when the tab is changed to decide to show or hide the banner or adjust the `safeAreaInset` dynamically on these visibility changes, and instead the ViewController associated with the 2 Jetpack-powered tabs (Reader and Notifications)  would get the banner for free without having to change anything in the implementation of `NotificationsViewController` and/or `StatsViewController` themselves.

But again, that's just an idea from someone who hasn't done iOS and UIKit in forever, and who hasn't a clue about how the architecture of the UI and ViewControllers currently works and is implemented in the current code, so I might very well be missing something or not be aware of edge cases which would make that suggestion not a good fit here… Still, figured I'd drop the idea anyway as my 2 cents it case that helps 😛 😅

**Comment by guarani** [comment]
I took the liberty of trying a different approach for the Stats screen where I added a view to its existing container view controller. The result is that the banner only shows on the Stats screen and the scrolling content is inset correctly:

https://user-images.githubusercontent.com/1898325/176964622-232e91e7-45ae-4676-93aa-5a6055972253.mov

This approach leverages the existing code to create the banner but removes the banner's association with the tab bar controller. I think a similar approach could be used for Notifications, which due to its tabbed structure, likely already uses a container view controller. Update: Notifications doesn't use a container view controller yet, but we could still add one perhaps.

Wdyt @Gio2018? My hope is that this alternative approach makes things simpler. It should also keep the banner within the detail view controller on the iPad, so it won't extend all the way across the width of the iPad (and thus no longer conflict with the FAB button).

**Comment by guarani** [comment]
I make a PR to demo this here: https://github.com/wordpress-mobile/WordPress-iOS/pull/18978

**Comment by Gio2018** [comment]
Thank you for the comments and reviews @AliSoftware  and @Guarani !!

>  I'm wondering if it wouldn't be simpler and also safer to instead: Make a [Container ViewController](https://developer.apple.com/library/archive/featuredarticles/ViewControllerPGforiPhoneOS/ImplementingaContainerViewController.html) that [...]

it probably could, but then wouldn't we have the banner only on the detail section, on iPad? Which is probably acceptable? Not sure and I will look into it, but I also want to be mindful that this was requested to be executed quickly. Also, this is only to advertise Jetpack on the WP app, and will eventually be removed.

> Is it intentional that you use an UIButton… but don't associate a target/action to it?

Yes, the target will be added at a later time because the banner will be actionable

File: WordPress/Classes/ViewRelated/Jetpack Branding/WPTabBarController+JetpackBanner.swift (hunk 1 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Jetpack Branding/WPTabBarController+JetpackBanner.swift]: Nope, removed!

**Comment by AliSoftware** [comment]
Pinging @mokagio as this targets the frozen, release branch (especially on the second week of the beta-testing phase)

**Comment by mokagio** [comment]
@Gio2018 I noticed this PR now has conflicts with the release branch.

The clock's ticking for the build finalization. Have you considered moving this against `trunk`?

If I understand correctly, the impetus to target 20.2 was to get started sooner rather than later with making it clear which features come from Jetpack and which from vanilla WordPress. On that front, we already have https://github.com/wordpress-mobile/WordPress-iOS/pull/18978 on the release branch.

**Comment by guarani** [comment]
👋 @mokagio, I think this draft PR will be closed in favor of:
- https://github.com/wordpress-mobile/WordPress-iOS/pull/18978
- https://github.com/wordpress-mobile/WordPress-iOS/pull/18990

cc @Gio2018

**Comment by Gio2018** [comment]
> @Gio2018 I noticed this PR now has conflicts with the release branch.
> 
> The clock's ticking for the build finalization. Have you considered moving this against `trunk`?
> 
> If I understand correctly, the impetus to target 20.2 was to get started sooner rather than later with making it clear which features come from Jetpack and which from vanilla WordPress. On that front, we already have #18978 on the release branch.

Thanks @mokagio , I'm actually closing this one

**Pull request closed without merging**